csharp/ql/test/library-tests/ir/offbyone/OffByOneRA.ql

```diff
@@ -0,0 +1,65 @@
+import csharp
+import semmle.code.csharp.ir.IR
+import semmle.code.csharp.ir.rangeanalysis.RangeAnalysis
+
+/**
+ * Holds if the index expression of `aa` is less than or equal to the array length plus `k`.
+ */
+predicate boundedArrayAccess(ElementAccess aa, int k) {
+  exists(Instruction index, Instruction usage, Bound b, int delta |
+    (
+      // indexer access
+      usage.(CallInstruction).getAST() = aa
+      or
+      // array access
+      usage.(PointerAddInstruction).getAST() = aa
+    ) and
+    usage.getAnOperand().getDef() = index and
+    boundedInstruction(index, b, delta, true, _)
+  |
+    exists(PropertyAccess pa |
+      k = delta and
+      b.getInstruction().getAST() = pa and
+      pa.getProperty().getName() = "Length" and
+      pa.(QualifiableExpr).getQualifier().(VariableAccess).getTarget() = aa
+            .getQualifier()
+            .(VariableAccess)
+            .getTarget()
+    )
+    or
+    exists(ArrayCreation ac |
+      ac.getParent().(LocalVariableDeclExpr).getVariable() = aa
+            .getQualifier()
+            .(VariableAccess)
+            .getTarget() and
+      b instanceof ZeroBound and
+      if exists(ac.getLengthArgument(0))
```

Comment:
Can we refactor this to use `getArrayDim` ?

Reply:
Fixed in a previous comment